crates/bevy_gilrs/src/rumble.rs

```diff
@@ -0,0 +1,99 @@
+//! Handle user specified Rumble request events.  use crate::converter::convert_gamepad_id; use bevy_app::EventReader; use bevy_core::Time;
+use bevy_ecs::{
+    prelude::{EventReader, Res},
+    system::NonSendMut,
+};
+use bevy_input::gamepad::{GamepadRumbleRequest, RumbleIntensity};
+use bevy_log::{debug, warn};
+use bevy_time::Time;
+use bevy_utils::HashMap;
+use gilrs::{ff, GamepadId, Gilrs};
+
+use crate::converter::convert_gamepad_id;
+
+fn get_effect_type(intensity: RumbleIntensity) -> ff::BaseEffectType {
+    use RumbleIntensity::*;
+    match intensity {
+        Strong => ff::BaseEffectType::Strong { magnitude: 63_000 },
+        Medium => ff::BaseEffectType::Strong { magnitude: 40_000 },
+        Weak => ff::BaseEffectType::Weak { magnitude: 40_000 },
```

Comment:
@alice-i-cecile raised a concern in the previous PR was that these values seemed somewhat arbitrary.

@nicopap do you remember why you chose these values specifically? Is there another less arbitrary scale we could use?

Reply:
It's fully arbitrary. I picked them based on how it felts in my hand with my controller (a logitech F710).

Reply:
I did some digging.

- The scale in gilrs goes from 0 to `u16::MAX`, because that's what [xinput does](https://learn.microsoft.com/en-us/windows/win32/api/xinput/ns-xinput-xinput_vibration?source=recommendations)
- The `Strong` and `Weak` `BaseEffectType`s map to left or right motors in the windows API. On linux, they map to `strong_magnitude` and `weak_magnitude` on the `ff_rumble` ioctl, on wasm and macos, it does nothing.
- I *think* the convention is that the left motor in most controllers is a low frequency motor, while the right is a high frequency.
- I think It might make sense to spin both motors at full speed for maximum rumble.

So it seems that on the two platforms they don't perfectly agree on which motor is which, but by convention it kind of is the same... So I'm wondering how to present a unified, and easy to use API.

I'm thinking it would at least make sense to make the scale be an `f32` from 0.0 to 1.0. Using `u16` seems a bit too implementation-specific and low level to be convenient to use, even though both windows xinput and linux ioctls use it.

For how to deal with the separate motors and speeds, I see the following options:

1. Keep it almost as is for now, but increase the Strong level to `u16::MAX`.
2. Add an `enum RumbleMotor { Strong, Weak }` and expose it in the event.

I also think it probably makes sense to make the effects additive instead of overwriting, at least by default.

If we want to provide some way to cancel running rumbles, we should rather return some kind of handle instead. Perhaps it would make sense to have a `Rumbles` resource instead of using events?

Reply:
3. Remove intensity in this PR and just provide duration.

Reply:
checked a few other game engines. godot also uses separate "strong" and "weak" magnitudes that go from 0 to 1: https://docs.godotengine.org/en/stable/classes/class_input.html#class-input-method-start-joy-vibration

Reply:
Awesome! Thanks for taking the time to clear things up.  I think the gilrs API is very weird (as you uncovered here!) and we should strive to wrap it into something sensible.

When making a rumble API, I see two questions to answer:

1. How should the user "talk" with the rumble system?
2. How to manage the "high frequency" and "low frequency" motors?

### 1. Talking with gamepads

I see three options:

- **gamepads as entities**: Gamepad state (including rumble) could be encoded as components. This also makes it easier to interact with gamepads in general.
- **rumble resource**: the least controversial API would be to add a new resource containing a `Vec<Rumble>`, (or `HashMap<GamepadId, Rubmle>`) where `Rumble` has a single field `intensity: f32`. The `RumbleManagers` would be updated by a system that reads the rumble resource.
- **event API**:

```rust
enum RumbleRequest {
    Start { gamepad: GamepadId, duration: f32, intensity: f32 },
    Stop { gamepad: GamepadId },
}
```

### 2. Specifying motors

I think the API should either forego individual motor controls or be clear about what `Strong` and `Weak` mean. Those are very misleading names, since they don't describe what they represent. They should be clearly documented in a `GamepadRumbleMotor` enum, or fields of whatever resource/component we chose to add.

---

Personally I'd go with the resource. This way the question of "overwrite vs additive" solves itself and it's possible for the user to inspect the rumble state before modifying it. Also it's a much less controversial change than gamepads as entities, even though I think that would be the most sensible solution.

It's still technically more challenging, since you have to convert it into the gilrs representation based on durations. It's also a bit harder to come up with a good internal representation.

Also yes please `f32: [0, 1]` is so much better than a `u16`, I think this is not controversial.

Reply:
> ## 2. Specifying motors

I focused on this first: While I also kind of felt like it would have been better to have a single strength value, the problem that then presents itself, how do we map that to the two motor values exposed in the underlying API?

Do we set both motors to the same `intensity`, or would it make more sense to scale the weak motor to max before we even start enabling the strong one? I opted to expose both values, at least for now, since that's what both other engines and the lower-level APIs seem to do.

> ## 1. Talking with gamepads

I was thinking a bit about how and when you would use it. One very common case would that the character just got hit by something, fell to the ground etc. In those cases, what you want is a fire and forget style API (rumble at this intensity for this many seconds). Also, you'd probably want them to add up, you don't want a long low intensity rumble to cancel an ongoing "important" rumble (like a character losing health).

Same goes for ui buttons, you're probably mostly interested to set a given strength and duration, then forget about it.

I'm a bit uncertain how to implement that in a resource-style API... We could perhaps do something like the following?

```rust
// bevy_input
#[derive(Resource)]
pub struct GamepadRumbles(HashMap<Gamepad, Vec<GamepadRumble>>);

impl GamepadRumbles {
    fn stop_all()
    fn stop(gamepad: Gamepad)
    fn add_rumble(...)
    fn replace_rumble()
}

struct GamepadRumble {
    deadline: f32
    instensity: GamepadRumbleIntensity,
}
```

And then do book-keeping in bevy_gilrs to sync the state?

Reply:
I went with events because that was how I implemented it in my game before opening the PR. So it makes sense to keep it that way. A case can be made that both modes of control could co-exist (at least for the user to be able to read the current rumble state), but I think IMO this PR could be merged even with only the event API.

With regard to individual motor controls based on events, you could have distinct events to send for the high magnitude and the low magnitude motors. But I think we should have a similar API to that of godot you linked, by having two intensity fields in the `Start` event. `hight_magnitude_motor_intensity: f32` and `low_magnitude_motor_intensity: f32` (of course, names could/should be different)

You suggestion that rumble intensity from events should be additive is growing on me even though I was initially doubtful. It solves a usability problem fairly cleanly. But I think this needs to be documented.

Reply:
Okay, so to sum up we keep this PR mostly as is? Then perhaps revisit adding a resource in another PR?

Will add some inline comments in the latest diff with my remaining questions to keep the discussions more narrow and easier to follow.

Reply:
Agreed with leaving this mostly as is. Intensity control is a must-have IMO.